google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java

```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
-        && Objects.equals(leaderOptions, that.leaderOptions);
+        && Objects.equals(leaderOptions, that.leaderOptions)
+        && Objects.equals(optionalReplicas, that.optionalReplicas)
+        && Objects.equals(baseConfig, that.baseConfig)
+        && Objects.equals(configType, that.configType)
+        && Objects.equals(etag, that.etag)
+        && Objects.equals(reconciling, that.reconciling)
+        && Objects.equals(state, that.state)
+        && Objects.equals(labels, that.labels);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(id, displayName, replicas, leaderOptions);
+    return Objects.hash(
+        id,
+        displayName,
+        replicas,
+        leaderOptions,
+        optionalReplicas,
+        baseConfig,
+        configType,
+        etag,
+        reconciling,
+        state,
+        labels);
+  }
+
+  public Builder toBuilder() {
+    return new BuilderImpl(this);
   }
 
   @Override
   public String toString() {
     return String.format(
-        "Instance Config[%s, %s, %s, %s]", id, displayName, replicas, leaderOptions);
+        "Instance Config[%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s]",
+        id,
+        displayName,
+        replicas,
+        leaderOptions,
+        optionalReplicas,
+        baseConfig,
+        configType,
+        etag,
+        reconciling,
+        state,
+        labels);
+  }
+
+  com.google.spanner.admin.instance.v1.InstanceConfig toProto() {
+    InstanceConfig.Builder builder =
+        com.google.spanner.admin.instance.v1.InstanceConfig.newBuilder()
+            .setName(getId().getName())
+            .setDisplayName(getDisplayName())
+            .addAllReplicas(
+                getReplicas().stream().map(ReplicaInfo::getProto).collect(Collectors.toList()))
+            .addAllLeaderOptions(getLeaderOptions())
+            .setEtag(getEtag())
+            .setReconciling(getReconciling())
+            .putAllLabels(getLabels())
+            .addAllOptionalReplicas(
+                getOptionalReplicas().stream()
+                    .map(ReplicaInfo::getProto)
+                    .collect(Collectors.toList()))
+            .setConfigType(toProtoConfigType(getConfigType()))
+            .setState(toProtoState(getState()));
+
+    if (getBaseConfig() != null) {
+      builder.setBaseConfig(getBaseConfig().getId().getName());
+    }
+
+    return builder.build();
+  }
+
+  static InstanceConfig.Type toProtoConfigType(Type type) {
+    switch (type) {
+      case TYPE_UNSPECIFIED:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.Type.TYPE_UNSPECIFIED;
+      case GOOGLE_MANAGED:
+        return com.google.spanner.admin.instance.v1.InstanceConfig.Type.GOOGLE_MANAGED;
+      case USER_MANAGED:
+        return InstanceConfig.Type.USER_MANAGED;
+      default:
+        throw new IllegalArgumentException("Unknown config type:" + type);
+    }
+  }
+
+  static InstanceConfig.State toProtoState(State state) {
```

Comment:
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

Reply:
Making it private makes sense. It is not directly used in unit tests, so I think we need not add @VisibleForTesting annotation.